Clang-tidy:
```
Jul 03 07:20:18 /build/build_docker/../src/Disks/ObjectStorages/S3/registerDiskS3.cpp:16:1: error: duplicate include [readability-duplicate-include,-warnings-as-errors]
```

More importantly, there are some S3 test failures which look related. Would you like to check? Thanks~

--- a/src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp
+++ b/src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp
@@ -228,32 +228,12 @@ void S3ObjectStorage::removeObjectImpl(const std::string & path, bool if_exists)
 {
     auto client_ptr = client.get();
 
-    // If chunk size is 0, only use single delete request
-    // This allows us to work with GCS, which doesn't support DeleteObjects
-    if (!s3_capabilities.support_batch_delete)
-    {
-        Aws::S3::Model::DeleteObjectRequest request;
-        request.SetBucket(bucket);
-        request.SetKey(path);
-        auto outcome = client_ptr->DeleteObject(request);
+    Aws::S3::Model::DeleteObjectRequest request;
+    request.SetBucket(bucket);
+    request.SetKey(path);
+    auto outcome = client_ptr->DeleteObject(request);
 
-        throwIfUnexpectedError(outcome, if_exists);
-    }
-    else
-    {
-        /// TODO: For AWS we prefer to use multiobject operation even for single object
-        /// maybe we shouldn't?
-        Aws::S3::Model::ObjectIdentifier obj;
-        obj.SetKey(path);
-        Aws::S3::Model::Delete delkeys;
-        delkeys.SetObjects({obj});
-        Aws::S3::Model::DeleteObjectsRequest request;
-        request.SetBucket(bucket);
-        request.SetDelete(delkeys);
-        auto outcome = client_ptr->DeleteObjects(request);
-
-        throwIfUnexpectedError(outcome, if_exists);
-    }
+    throwIfUnexpectedError(outcome, if_exists);
 }
 
 void S3ObjectStorage::removeObjectsImpl(const PathsWithSize & paths, bool if_exists)

--- a/src/Disks/ObjectStorages/S3/S3ObjectStorage.h
+++ b/src/Disks/ObjectStorages/S3/S3ObjectStorage.h
@@ -83,12 +83,18 @@ public:
     void listPrefix(const std::string & path, RelativePathsWithSize & children) const override;
 
     /// Remove file. Throws exception if file doesn't exist or it's a directory.
+    /// Uses `DeleteObjectRequest`.
     void removeObject(const std::string & path) override;
 
+    /// Uses `DeleteObjectsRequest` if it is allowed by `s3_capabilities`, otherwise `DeleteObjectRequest`.
+    /// `DeleteObjectsRequest` is not supported on GCS, see https://issuetracker.google.com/issues/162653700 .
     void removeObjects(const PathsWithSize & paths) override;
 
+    /// Uses `DeleteObjectRequest`.
     void removeObjectIfExists(const std::string & path) override;
 
+    /// Uses `DeleteObjectsRequest` if it is allowed by `s3_capabilities`, otherwise `DeleteObjectRequest`.
+    /// `DeleteObjectsRequest` does not exist on GCS, see https://issuetracker.google.com/issues/162653700 .
     void removeObjectsIfExist(const PathsWithSize & paths) override;
 
     ObjectMetadata getObjectMetadata(const std::string & path) const override;
@@ -112,6 +118,8 @@ public:
         const Poco::Util::AbstractConfiguration & config,
         const std::string & config_prefix,
         ContextPtr context) override;
+
+    void setCapabilitiesSupportBatchDelete(bool value) { s3_capabilities.support_batch_delete = value; }
 
     String getObjectsNamespace() const override { return bucket; }
 
@@ -151,7 +159,7 @@ private:
 
     MultiVersion<Aws::S3::S3Client> client;
     MultiVersion<S3ObjectStorageSettings> s3_settings;
-    const S3Capabilities s3_capabilities;
+    S3Capabilities s3_capabilities;
 
     const String version_id;
 };

--- a/src/Disks/ObjectStorages/S3/registerDiskS3.cpp
+++ b/src/Disks/ObjectStorages/S3/registerDiskS3.cpp
@@ -11,12 +11,14 @@
 #include <aws/core/client/DefaultRetryStrategy.h>
 
 #include <base/getFQDNOrHostName.h>
+
 #include <Common/FileCacheFactory.h>
+
+#include <IO/S3Common.h>
 
 #include <Disks/DiskCacheWrapper.h>
 #include <Disks/DiskRestartProxy.h>
 #include <Disks/DiskLocal.h>
-
 #include <Disks/ObjectStorages/DiskObjectStorage.h>
 #include <Disks/ObjectStorages/DiskObjectStorageCommon.h>
 #include <Disks/ObjectStorages/S3/ProxyConfiguration.h>
@@ -26,9 +28,8 @@
 #include <Disks/ObjectStorages/S3/diskSettings.h>
 #include <Disks/ObjectStorages/MetadataStorageFromDisk.h>
 
-#include <IO/S3Common.h>
-
 #include <Storages/StorageS3Settings.h>
+
 
 namespace DB
 {
@@ -65,7 +66,35 @@ void checkReadAccess(const String & disk_name, IDisk & disk)
         throw Exception("No read access to S3 bucket in disk " + disk_name, ErrorCodes::PATH_ACCESS_DENIED);
 }
 
-void checkRemoveAccess(IDisk & disk) { disk.removeFile("test_acl"); }
+void checkRemoveAccess(IDisk & disk)
+{
+    disk.removeFile("test_acl");
+}
+
+bool checkBatchRemoveIsMissing(S3ObjectStorage & storage, const String & key)
+{
+    String path = key + "/test_acl";
+    auto file = storage.writeObject(path, WriteMode::Rewrite);
+    try
+    {
+        file->write("test", 4);
+    }
+    catch (...)
+    {
+        file->finalize();
+        return false; /// We don't have write access, therefore no information about batch remove.
+    }
+    try
+    {
+        /// Uses `DeleteObjects` request (batch delete).
+        storage.removeObjects({{ path, 0 }});
+        return false;
+    }
+    catch (const Exception &)
+    {
+        return true;
+    }
+}
 
 }
 
@@ -91,10 +120,26 @@ void registerDiskS3(DiskFactory & factory)
         FileCachePtr cache = getCachePtrForDisk(name, config, config_prefix, context);
         S3Capabilities s3_capabilities = getCapabilitiesFromConfig(config, config_prefix);
 
-        ObjectStoragePtr s3_storage = std::make_unique<S3ObjectStorage>(
+        auto s3_storage = std::make_unique<S3ObjectStorage>(
             std::move(cache), getClient(config, config_prefix, context),
             getSettings(config, config_prefix, context),
             uri.version_id, s3_capabilities, uri.bucket);
+
+        if (!config.getBool(config_prefix + ".skip_access_check", false))
+        {
+            /// If `support_batch_delete` is turned on (default), check and possibly switch it off.
+            if (s3_capabilities.support_batch_delete && checkBatchRemoveIsMissing(*s3_storage, uri.key))
+            {
+                LOG_WARNING(
+                    &Poco::Logger::get("registerDiskS3"),
+                    "Storage for disk {} does not support batch delete operations, "
+                    "so `s3_capabilities.support_batch_delete` was automatically turned off during the access check. "
+                    "To remove this message set `s3_capabilities.support_batch_delete` for the disk to `false`.",
+                    name
+                );
+                s3_storage->setCapabilitiesSupportBatchDelete(false);
+            }
+        }
 
         bool send_metadata = config.getBool(config_prefix + ".send_metadata", false);
         uint64_t copy_thread_pool_size = config.getUInt(config_prefix + ".thread_pool_size", 16);

